main/src/ca/uwaterloo/flix/language/phase/Parser2.scala

```diff
@@ -2014,58 +2025,72 @@ object Parser2 {
-            case TokenKind.ParenR => parenNestingLevel -= 1; lookAhead += 1
-            case TokenKind.Eof =>
-              val error = UnexpectedToken(expected = NamedTokenSet.Expression, actual = None, sctx, loc = currentSourceLocation())
-              return closeWithError(mark, error)
-            case t if t.isFirstDecl =>
-              // Advance past the erroneous region to the next stable token
-              // (the start of the declaration).
-              for (_ <- 0 until lookAhead) {
-                advance()
-              }
-              val error = UnexpectedToken(expected = NamedTokenSet.Expression, actual = Some(t), sctx, loc = currentSourceLocation())
-              return closeWithError(mark, error)
-            case _ => lookAhead += 1
+      isMatchLambda(mark) match {
+        case Result.Err(errMark) => errMark
+        case Result.Ok(isLambda) =>
+          if (isLambda) {
+            Pattern.pattern()
+            expect(TokenKind.ArrowThinR)
+            expression()
+            close(mark, TreeKind.Expr.LambdaMatch)
+          } else {
+            expression()
+            oneOrMore(
+              namedTokenSet = NamedTokenSet.MatchRule,
+              checkForItem = _ == TokenKind.KeywordCase,
+              getItem = matchRule,
+              breakWhen = _.isRecoverExpr,
+              delimiterL = TokenKind.CurlyL,
+              delimiterR = TokenKind.CurlyR,
+              separation = Separation.Optional(TokenKind.Comma)
+            )
+            close(mark, TreeKind.Expr.Match)
           }
-        }
-        isLambda
       }
+    }
 
-      if (isLambda) {
-        Pattern.pattern()
-        expect(TokenKind.ArrowThinR)
-        expression()
-        close(mark, TreeKind.Expr.LambdaMatch)
-      } else {
-        expression()
-        oneOrMore(
-          namedTokenSet = NamedTokenSet.MatchRule,
-          checkForItem = _ == TokenKind.KeywordCase,
-          getItem = matchRule,
-          breakWhen = _.isRecoverExpr,
-          delimiterL = TokenKind.CurlyL,
-          delimiterR = TokenKind.CurlyR,
-          separation = Separation.Optional(TokenKind.Comma)
-        )
-        close(mark, TreeKind.Expr.Match)
+    /**
+      * Returns `Result.Ok(true)` if starting a (e)match-lambda, i.e., given a `match` or `ematch` keyword it appears in the position `(match ... -> ...)`.
+      * Returns `Result.Ok(false)` if this is not the case.
+      * Returns `Result.Err(closeWithError(mark))` if a parser error occurs.
+      *
+      * We use the `Result` type here to allow the caller to perform an early return in the case of a parser error.
+      *
+      * @param mark the mark opened before the `match` or `ematch` keyword, i.e., `(*mark* match ... -> ...)`.
+      *             Additionally, the caller must have consumed the `match` or `ematch` keyword, so the cursor is immediately after the keyword, i.e., `(match *cursor* ... -> ...)`.
+      */
```

Comment:
Maybe I am tired, but I struggle a little bit to understand the above. Like I partially understand it, but not fully.